sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetValidation.java

```diff
@@ -0,0 +1,292 @@
+  }
+
+  /**
+   * Checks if a buffer contains only valid UTF8 encoded bytes. 
+   * It's very effective, giving a clear YES/NO, unless it's ASCII  (unicode < 127), in which case it returns MAYBE.
+   * 
+   * 
+   * First byte:
+   * 0xxxxxxx: only one byte (0-127)
+   * 110xxxxx: 2 bytes       (194-223, as 192/193 are invalid)
+   * 1110xxxx: 3 bytes       (224-239)
+   * 11110xxx: 4 bytes       (240-244)
+   * 
+   * Bytes 2,3 and 4 are always 10xxxxxx (0x80-0xBF or 128-191).
+   * 
+   * So depending on the number of significant bits in the unicode code point, the length will be 1,2,3 or 4 bytes:
+   * 0 -7 bits  (0x0000-007F):  1 byte encoding
+   * 8 -11 bits (0x0080-07FF): 2 bytes encoding
+   * 12-16 bits (0x0800-FFFF): 3 bytes encoding
+   * 17-21 bits (0x10000-10FFFF): 4 bytes encoding
+   */
+  public Result isUTF8(byte[] buffer, boolean rejectNulls) {
+    boolean onlyAscii = true;
+
+    for (int i = 0; i < buffer.length; i++) {
+      byte len;
+      // make it unsigned for the comparisons
+      int c = (0xFF) & buffer[i];
+
+      if (rejectNulls && c == 0) {
+        return Result.newInvalid();
+      }
+      if (c <= 127) {
+        len = 0;
+      } else if (c >= 194 && c <= 223) {
+        len = 1;
+      } else if (c >= 224 && c <= 239) {
+        len = 2;
+      } else if (c >= 240 && c <= 244) {
+        len = 3;
+      } else {
+        return Result.newInvalid();
+      }
+
+      for (i++; len > 0 && i < buffer.length; i++) {
+        c = buffer[i];
+        onlyAscii = false;
+
+        // first 2 bits should be 10
+        if ((c & 0b11000000) != 0b10000000) {
+          return Result.newInvalid();
+        }
+        len--;
+      }
+    }
+
+    return onlyAscii ? new Result(Valid.MAYBE, StandardCharsets.UTF_8) : Result.newValid(StandardCharsets.UTF_8);
+  }
+
+  /**
+   * Tries to use the given charset to decode the byte array.
+   * @return true if decoding succeeded, false if there was a decoding error.
+   */
+  public boolean tryDecode(byte[] bytes, @Nullable Charset charset) {
+    if (charset == null) {
+      return false;
+    }
+    CharsetDecoder decoder = charset.newDecoder()
+      .onMalformedInput(CodingErrorAction.REPORT)
+      .onUnmappableCharacter(CodingErrorAction.REPORT);
+
+    try {
+      decoder.decode(ByteBuffer.wrap(bytes));
+    } catch (CharacterCodingException e) {
+      return false;
+    }
+    return true;
+  }
+
+  private int read16bit(byte[] buffer, int i, boolean le) {
+    return le ? (buffer[i / 2] & 0xff) | (buffer[i / 2 + 1] & 0xff) << 8
+      : (buffer[i / 2] & 0xff) << 8 | (buffer[i / 2 + 1] & 0xff);
+  }
+
+  public enum Valid {
+    NO,
+    YES,
+    MAYBE
+  }
+
+  public static class Result {
+    private Valid valid;
+    private Charset charset;
+
+    private Result(Valid valid, @Nullable Charset charset) {
+      this.valid = valid;
+      this.charset = charset;
+    }
+
+    public static Result newInvalid() {
+      return new Result(Valid.NO, null);
```

Comment:
Since this can be immutable, it can be a constant, no need for a new instance every time.